client/services/FeedbackGeneratorService.js

```diff
@@ -14,18 +14,47 @@
 
 /**
  * @fileoverview Service for generating feedback based on the evaluation result
- * of the code submitted by the user.
+ * of the code submitted by the user, preprocessed and augmented with unit
+ * tests. This assumes that the base code already passes "compile-time" checks.
  */
 
 tie.factory('FeedbackGeneratorService', [
   function() {
+    // TODO(sll): Add escaping?
+    var jsToHumanReadable = function(jsVariable) {
+      if (typeof jsVariable === 'string') {
+        return '"' + jsVariable + '"';
+      }
+    };
+
     return {
-      getFeedback: function(question, codeEvalResult) {
+      getFeedback: function(prompt, codeEvalResult) {
         if (codeEvalResult.getErrorMessage()) {
           return (
             'Your code threw an error: ' + codeEvalResult.getErrorMessage());
         } else {
-          return 'Your code gave this output: ' + codeEvalResult.getOutput();
+          var correctnessTests = prompt.getCorrectnessTests();
+          var observedOutputs = codeEvalResult.getTestResults();
+          for (var i = 0; i < correctnessTests.length; i++) {
+            var expectedOutput = correctnessTests[i].getExpectedOutput();
+            var observedOutput = observedOutputs[i];
+
+            if (expectedOutput !== observedOutput) {
+              return [
+                'Your code gave the output ',
```

Comment:
Good point, but let's defer that to a separate incremental PR? I think that's a nice chunk of work on its own, and I don't want this one to get too big.